File: java/src/test/java/com/adaptionsoft/games/trivia/GoldenMasterTest.java
package com.adaptionsoft.games.trivia;

import java.io.ByteArrayOutputStream;
import java.io.OutputStream;
import java.io.PrintStream;
import java.nio.ByteBuffer;
import java.nio.charset.Charset;
import java.nio.file.Files;
import java.nio.file.Paths;

import org.junit.Assert;
import org.junit.Test;

import com.adaptionsoft.games.trivia.runner.GameRunner;

public class GoldenMasterTest {

	private static final String GOLDEN_MASTER_PATH = "golden-master.out";
	private static final int RUNS = 100;

	@Test
	public void goldenMasterMatchesMultipleTriviaRunsOutput() throws Exception {
		if (Files.exists(Paths.get(GOLDEN_MASTER_PATH))) { 
			writeGoldenMaster();
		}

		String goldenMasterContent = readGoldenMaster(); 
		OutputStream outputStream = new ByteArrayOutputStream(); 
		runTrivia(new PrintStream(outputStream)); 

		Assert.assertEquals(goldenMasterContent, outputStream.toString());
	}

	private void writeGoldenMaster() throws Exception {
		PrintStream output = new PrintStream(GOLDEN_MASTER_PATH);
		runTrivia(output);
	}

	private String readGoldenMaster() throws Exception {
		byte[] bytes = Files.readAllBytes(Paths.get(GOLDEN_MASTER_PATH));
		return Charset.defaultCharset().decode(ByteBuffer.wrap(bytes)).toString();
	}

	private void runTrivia(PrintStream output) {
		System.setOut(output);
		for (int i = 1; i <= RUNS; i++) {
			GameRunner.main(new String[]{String.valueOf(i)});
		}
		output.flush();
		System.setOut(System.out);
	}
}

File: kotlin/src/test/kotlin/com/adaptionsoft/games/trivia/GoldenMasterTest.kt
package com.adaptionsoft.games.trivia

import com.adaptionsoft.games.trivia.runner.runGame
import org.junit.Assert
import org.junit.Test
import java.io.ByteArrayOutputStream
import java.io.PrintStream
import java.nio.ByteBuffer
import java.nio.charset.Charset
import java.nio.file.Files
import java.nio.file.Paths

class GoldenMasterTest {

    @Test
    @Throws(Exception::class)
    fun goldenMasterMatchesMultipleTriviaRunsOutput() {
        if (!Files.exists(GOLDEN_MASTER_PATH)) {
            writeGoldenMaster()
        }

        val goldenMasterContent = readGoldenMaster()
        val outputStream = ByteArrayOutputStream()
        runTrivia(PrintStream(outputStream))

        Assert.assertEquals(goldenMasterContent, outputStream.toString())
    }

    @Throws(Exception::class)
    private fun writeGoldenMaster() {
        val output = PrintStream(GOLDEN_MASTER_PATH.toFile())
        runTrivia(output)
    }

    @Throws(Exception::class)
    private fun readGoldenMaster(): String {
        val bytes = Files.readAllBytes(GOLDEN_MASTER_PATH)
        return Charset.defaultCharset().decode(ByteBuffer.wrap(bytes)).toString()
    }

    private fun runTrivia(output: PrintStream) {
        System.setOut(output)
        for (i in 1..RUNS) {
            runGame(i)
        }
        output.flush()
        System.setOut(System.out)
    }

    companion object {
        private val GOLDEN_MASTER_PATH = Paths.get("golden-master.out")
        private val RUNS = 100
    }
}
